### ISSUE TITLE
Define the coordinate system

### ISSUE THREAD

**Issue by almarklein** [pygfx/pygfx#460]
We should:
* Decide on a coordinate system
* Document it

### What do others use?

**pygfx currently** we seem to have a right-handed system with right +X, up +Y and forward -Z.

**glTF** uses a right-handed coordinate system. glTF defines +Y as up, +Z as forward, and -X as right; the front of a glTF asset faces +Z. The units for all linear distances are meters. All angles are in radians. Positive rotation is counterclockwise.

**OpenGL** by convention uses a right-handed system. What this basically says is that the positive x-axis is to your right, the positive y-axis is up and the positive z-axis is backwards.

**Blender** [uses](http://www.rab3d.com/tut_blen_started.php) the [right hand](http://en.wikipedia.org/wiki/Right-hand_rule) coordinate system with the Z axis pointing upwards. This is common with the coordinate systems used by most common 3D CAD packages.

From my understanding, CAD software like 3D studio max and blender historically have Z point upwards, whereas game and gfx related software has Y point upwards. Not sure why glTF swaps directions for X and Z ...

**Comment by Korijn** [comment]
Can you clarify in each instance whether or not the coordinate system applies to model/view/clip space? Because the conventions are different between those spaces usually as well.

**Comment by FirefoxMetzger** [comment]
> pygfx currently we seem to have a right-handed system with right +X, up +Y and forward -Z.

😆 sigh, so which is it now, does `forward` mean positive Z or negative Z? 😅

**Comment by Korijn** [comment]
I found this nice picture: :)

![image](https://user-images.githubusercontent.com/1882046/219696037-28d75bd6-ffd8-4075-abd4-41fa69e07b36.png)

I would like to add that Three.js is Y-up, and right-handed.

Also, you can actually change the coordinate system in three like this:

`THREE.Object3D.DefaultUp = new THREE.Vector3(0,0,1);`

But that's all besides the point. We just need to pick something and document it. What do we want?

**Comment by almarklein** [comment]
> 😆 sigh, so which is it now, does forward mean positive Z or negative Z? 😅

Sorry, I know! I was confused between NDC and model coords! 😆 

To quote Linkin Park: In the end, it doesn't even matter (what we choose). Users can always set the camera's up value, and/or flip the camera's scale to move from a left-handed to right-handed system. The main issue (as I understand it) is that when a user loads left-handed models into a right-handed system (or vise versa) the models will be inside-out. And that only is really a problem when people set the `material.side = 'front'`.

> The Z negative/positive thing seems to be our only real uncertainty here. Let's pick positive Z?

From what I understand, given +X is right, and +Y is up, if we'd pick forward to mean +Z it'd be left-handed, and -Z would be right-handed.

**Comment by almarklein** [comment]
Addition: we're not actually deciding on *the* coordinate system, but only on the default.

**Comment by Korijn** [comment]
> From what I understand, given +X is right, and +Y is up, if we'd pick forward to mean +Z it'd be left-handed, and -Z would be right-handed.

You are right. And since we want to be right-handed, let's pick -Z as forward.

![image](https://user-images.githubusercontent.com/1882046/220059367-4513077a-f7d9-415b-86a5-992e4e71a06f.png)

That aligns with three.js defaults.

* +X is right
* +Y is up
* -Z is forward

Can we close this?

**Comment by almarklein** [comment]
👍

**Comment by Korijn** [comment]
I'm reopening this because I discovered some new information. Three.js seems to follow the glTF conventions.

I summarized them here: https://github.com/pygfx/pylinalg/pull/43#issuecomment-1438359082

I propose that we align with those conventions as well, both in pygfx and pylinalg.

**Comment by Korijn** [comment]
So to summarize from the glTF spec:

* glTF uses a right-handed coordinate system. glTF defines +Y as up, +Z as forward, and -X as right; the front of a glTF asset faces +Z.
* The camera is defined such that the local +X axis is to the right, the “lens” looks towards the local -Z axis, and the top of the camera is aligned with the local +Y axis.
The view matrix is derived from the global transform of the node containing the camera with the scaling ignored. If the node’s global transform is identity, the location of the camera is at the origin.
* For light types that have a direction (directional and spot lights), the light's direction is defined as the 3-vector (0.0, 0.0, -1.0) and the rotation of the node orients the light accordingly. That is, an untransformed light points down the -Z axis. The light's transform is affected by the node's world scale, but all properties of the light (such as range and intensity) are unaffected.

**Comment by almarklein** [comment]
I've copied this to the top post and also started a todo list. Feel free to add items.

**Comment by Korijn** [comment]
@almarklein can you comment on the status of this item? I think we've completed all the implementation changes, and just need to document the conventions now, right?

**Comment by almarklein** [comment]
I thought we already documented it too, but searching for "handed" in the docs shows nothing, so I guess we should add it. My suggestion would be to explain in in the guide, as well as a brief technical summary on the base camera.

**Comment by DennyLindberg** [comment]
Could I request documentation on linear distance units as well? I tried figuring out the default units and could not find any information on it. glTF is defined in meters ([see 3.4](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html)). The helmet in the PBR example is 1.5-2.0 meter sized when opened in Blender and that size is reflected in pygfx when using the trimesh python package. If I create a reference mesh in Blender and export to various formats using the default scene settings the size is consistent in pygfx to the meter standard. I get the impression that many of the pygfx examples are either using cm or unitless which can sometimes make it confusing when the camera setup is automatic for framing the object. I see values in the 20, 50, 100, 600 ranges just skimming through the examples now. Seeing that you are already documenting the coordinate system standards it would be nice if all units were clearly defined. I like pygfx so far so I would love for it to become a "standard" python package.

**Comment by almarklein** [comment]
@DennyLindberg I agree that documenting this in the docs is a good idea.

The short answer is that pygfx does not impose any units. It's the responsibility of the user to ensure that all objects are represented in the same units. In Blender-like use-case, and loading stuff using trimesh this would indeed often be meters. But for scientific applications it might be different. E.g. microscopy vs GIS vs astronomy.

**Comment by DennyLindberg** [comment]
@almarklein Fair enough.

**Comment by DennyLindberg** [comment]
Yes, that seems to be the same notion over at ThreeJS. It would not hurt to mention it is unitless right, if other details are being documented anyway?

https://discourse.threejs.org/t/what-unit-of-distance-in-threejs/20782/2
https://discourse.threejs.org/t/units-in-three-js/1174
https://github.com/mrdoob/three.js/issues/6259

**Comment by almarklein** [comment]
I added a checkbox in the top post.
